Clean up UI process back/forward state during provisional navigations

charliewolfe opened the pull request:
#### f51a9069c20e133bcf8b29b45fe56302b783eed8
<pre>
Clean up UI process back/forward state during provisional navigations
<a href="https://bugs.webkit.org/show_bug.cgi?id=281704">https://bugs.webkit.org/show_bug.cgi?id=281704</a>
<a href="https://rdar.apple.com/138150829">rdar://138150829</a>

Reviewed by Alex Christensen.

When a back/forward navigation begins, we update the back/forward list in the UI process before the
navigation commits. This allows a client to quickly navigate back or forward again, even before the
navigation commits. Because of this, we have logic in `FrameLoader::continueLoadAfterNavigationPolicy`
to return the back/forward cursor to the previous item if the provisional navigation was canceled.
Instead of doing this, we should make the list in the UI process aware when a back/forward navigation has
started but not yet committed. The back/forward list will then use the provisional cursor until it is
either cleared or notified to switch to a non-provisional item.

`HistoryItem` now no longer needs to track whether it belongs to a navigated frame, so `m_isTargetItem`
can be removed.

I also added a test to ensure the back/forward cursor still returns to the previous item after canceling
a provisional navigation.

* LayoutTests/contentfiltering/allow-after-add-data-expected.txt:
* LayoutTests/contentfiltering/allow-after-finished-adding-data-expected.txt:
* LayoutTests/contentfiltering/allow-after-response-expected.txt:
* LayoutTests/contentfiltering/allow-after-will-send-request-expected.txt:
* LayoutTests/contentfiltering/allow-never-expected.txt:
* LayoutTests/contentfiltering/block-after-add-data-expected.txt:
* LayoutTests/contentfiltering/block-after-add-data-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-add-data-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-response-expected.txt:
* LayoutTests/contentfiltering/block-after-response-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-response-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-never-expected.txt:
* LayoutTests/fast/history/back-from-page-with-focused-iframe-expected.txt:
* LayoutTests/fast/history/form-submit-in-frame-expected.txt:
* LayoutTests/fast/history/form-submit-in-frame-via-onclick-expected.txt:
* LayoutTests/fast/history/gesture-before-onload-form-submit-expected.txt:
* LayoutTests/fast/history/gesture-before-onload-location-href-expected.txt:
* LayoutTests/fast/history/history-back-initial-vs-final-url-expected.txt:
* LayoutTests/fast/history/redirect-via-iframe-expected.txt:
* LayoutTests/fast/history/same-document-iframes-changing-fragment-expected.txt:
* LayoutTests/fast/history/same-document-iframes-changing-pushstate-expected.txt:
* LayoutTests/fast/loader/dynamic-iframe-extra-back-forward-item-expected.txt:
* LayoutTests/fast/loader/form-submission-before-load-get-expected.txt:
* LayoutTests/fast/loader/form-submission-before-load-post-expected.txt:
* LayoutTests/fast/loader/fragment-after-redirect-gets-back-entry-expected.txt:
* LayoutTests/fast/loader/frame-location-change-not-added-to-history-expected.txt:
* LayoutTests/fast/loader/frame-src-change-added-to-history-expected.txt:
* LayoutTests/fast/loader/frame-src-change-not-added-to-history-expected.txt:
* LayoutTests/fast/loader/stateobjects/pushstate-clears-forward-history-expected.txt:
* LayoutTests/fast/loader/subframe-navigate-during-main-frame-load-expected.txt:
* LayoutTests/http/tests/history/back-with-fragment-change-expected.txt:
* LayoutTests/http/tests/history/history-replace-updates-current-item-expected.txt:
* LayoutTests/http/tests/history/redirect-200-refresh-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-200-refresh-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-301-expected.txt:
* LayoutTests/http/tests/history/redirect-302-expected.txt:
* LayoutTests/http/tests/history/redirect-303-expected.txt:
* LayoutTests/http/tests/history/redirect-307-expected.txt:
* LayoutTests/http/tests/history/redirect-js-document-location-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-document-location-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-document-location-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-js-form-submit-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-form-submit-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-form-submit-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-assign-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-assign-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-assign-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-href-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-href-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-href-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-replace-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-replace-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-replace-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-meta-refresh-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-meta-refresh-2-seconds-expected.txt:
* LayoutTests/http/tests/navigation/anchor-basic-expected.txt:
* LayoutTests/http/tests/navigation/anchor-goback-expected.txt:
* LayoutTests/http/tests/navigation/anchor-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/back-to-slow-frame-expected.txt:
* LayoutTests/http/tests/navigation/back-twice-without-commit-expected.txt:
* LayoutTests/http/tests/navigation/document-location-click-expected.txt:
* LayoutTests/http/tests/navigation/document-location-click-timeout-expected.txt:
* LayoutTests/http/tests/navigation/document-location-mouseover-expected.txt:
* LayoutTests/http/tests/navigation/document-location-onload-expected.txt:
* LayoutTests/http/tests/navigation/document-location-script-expected.txt:
* LayoutTests/http/tests/navigation/dynamic-iframe-dynamic-form-back-entry-expected.txt:
* LayoutTests/http/tests/navigation/error404-frames-expected.txt:
* LayoutTests/http/tests/navigation/forward-and-cancel-expected.txt:
* LayoutTests/http/tests/navigation/javascriptlink-basic-expected.txt:
* LayoutTests/http/tests/navigation/javascriptlink-goback-expected.txt:
* LayoutTests/http/tests/navigation/javascriptlink-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/location-assign-adds-history-item-expected.txt:
* LayoutTests/http/tests/navigation/location-href-set-adds-history-item-expected.txt:
* LayoutTests/http/tests/navigation/location-replace-adds-history-item-expected.txt:
* LayoutTests/http/tests/navigation/location-set-adds-history-item-expected.txt:
* LayoutTests/http/tests/navigation/lockedhistory-iframe-expected.txt:
* LayoutTests/http/tests/navigation/metaredirect-basic-expected.txt:
* LayoutTests/http/tests/navigation/metaredirect-frames-expected.txt:
* LayoutTests/http/tests/navigation/metaredirect-goback-expected.txt:
* LayoutTests/http/tests/navigation/metaredirect-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/multiple-back-forward-entries-expected.txt:
* LayoutTests/http/tests/navigation/navigation-interrupted-by-fragment-expected.txt:
* LayoutTests/http/tests/navigation/new-window-redirect-history-expected.txt:
* LayoutTests/http/tests/navigation/onload-navigation-iframe-2-expected.txt:
* LayoutTests/http/tests/navigation/onload-navigation-iframe-expected.txt:
* LayoutTests/http/tests/navigation/onload-navigation-iframe-timeout-expected.txt:
* LayoutTests/http/tests/navigation/parsed-iframe-dynamic-form-back-entry-expected.txt:
* LayoutTests/http/tests/navigation/post-basic-expected.txt:
* LayoutTests/http/tests/navigation/post-frames-expected.txt:
* LayoutTests/http/tests/navigation/post-frames-goback1-expected.txt:
* LayoutTests/http/tests/navigation/post-frames-goback1-uncached-expected.txt:
* LayoutTests/http/tests/navigation/post-goback-same-url-expected.txt:
* LayoutTests/http/tests/navigation/post-goback1-expected.txt:
* LayoutTests/http/tests/navigation/post-goback2-expected.txt:
* LayoutTests/http/tests/navigation/postredirect-frames-expected.txt:
* LayoutTests/http/tests/navigation/postredirect-frames-goback1-expected.txt:
* LayoutTests/http/tests/navigation/postredirect-goback2-expected.txt:
* LayoutTests/http/tests/navigation/postredirect-reload-expected.txt:
* LayoutTests/http/tests/navigation/redirect-cycle-expected.txt:
* LayoutTests/http/tests/navigation/redirect-load-no-form-restoration-expected.txt:
* LayoutTests/http/tests/navigation/redirect-on-back-updates-history-item-expected.txt:
* LayoutTests/http/tests/navigation/redirect-on-reload-updates-history-item-expected.txt:
* LayoutTests/http/tests/navigation/redirect302-basic-expected.txt:
* LayoutTests/http/tests/navigation/redirect302-frames-expected.txt:
* LayoutTests/http/tests/navigation/redirect302-goback-expected.txt:
* LayoutTests/http/tests/navigation/redirect302-metaredirect-expected.txt:
* LayoutTests/http/tests/navigation/redirect302-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/relativeanchor-basic-expected.txt:
* LayoutTests/http/tests/navigation/relativeanchor-frames-expected.txt:
* LayoutTests/http/tests/navigation/relativeanchor-goback-expected.txt:
* LayoutTests/http/tests/navigation/response204-expected.txt:
* LayoutTests/http/tests/navigation/restore-form-state-https-expected.txt:
* LayoutTests/http/tests/navigation/sec-fetch-site-header-expected.txt:
* LayoutTests/http/tests/navigation/sec-fetch-site-header-on-crossorigin-redirection-expected.txt:
* LayoutTests/http/tests/navigation/sec-fetch-site-header-on-redirection-expected.txt:
* LayoutTests/http/tests/navigation/slowmetaredirect-basic-expected.txt:
* LayoutTests/http/tests/navigation/slowtimerredirect-basic-expected.txt:
* LayoutTests/http/tests/navigation/statistics-expected.txt:
* LayoutTests/http/tests/navigation/success200-basic-expected.txt:
* LayoutTests/http/tests/navigation/success200-frames-expected.txt:
* LayoutTests/http/tests/navigation/success200-frames-loadsame-expected.txt:
* LayoutTests/http/tests/navigation/success200-goback-expected.txt:
* LayoutTests/http/tests/navigation/success200-loadsame-expected.txt:
* LayoutTests/http/tests/navigation/success200-reload-expected.txt:
* LayoutTests/http/tests/navigation/success200-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/timerredirect-basic-expected.txt:
* LayoutTests/http/tests/navigation/timerredirect-frames-expected.txt:
* LayoutTests/http/tests/navigation/timerredirect-goback-expected.txt:
* LayoutTests/http/tests/navigation/timerredirect-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/window-open-adds-history-item-expected.txt:
* LayoutTests/http/tests/navigation/window-open-adds-history-item2-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-allowed-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-default-ignored-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-get-allowed-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-get-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-javascript-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-self-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/upgrade-insecure-requests/proper-uir-on-navigation-expected.txt:
* LayoutTests/platform/glib/http/tests/navigation/error404-basic-expected.txt:
* LayoutTests/platform/glib/http/tests/navigation/error404-goback-expected.txt:
* LayoutTests/platform/glib/http/tests/navigation/postredirect-basic-expected.txt:
* LayoutTests/platform/glib/http/tests/navigation/postredirect-goback1-expected.txt:
* LayoutTests/platform/gtk/http/tests/navigation/error404-subframeload-expected.txt:
* LayoutTests/platform/gtk/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/error404-basic-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/error404-goback-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/error404-subframeload-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/postredirect-basic-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/postredirect-goback1-expected.txt:
* LayoutTests/platform/mac-ventura/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/mac-wk1/http/tests/navigation/sec-fetch-site-header-expected.txt:
* LayoutTests/platform/mac-wk1/http/tests/navigation/sec-fetch-site-header-on-crossorigin-redirection-expected.txt:
* LayoutTests/platform/mac-wk1/http/tests/navigation/sec-fetch-site-header-on-redirection-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/error404-basic-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/error404-goback-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/error404-subframeload-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/postredirect-basic-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/postredirect-goback1-expected.txt:
* LayoutTests/platform/wpe/http/tests/navigation/javascriptlink-frames-expected.txt:
* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::setProvisionalItem):
(WebCore::BackForwardController::clearProvisionalItem):
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::reset):
(WebCore::HistoryItem::setChildItem):
(WebCore::HistoryItem::isTargetItem const): Deleted.
(WebCore::HistoryItem::setIsTargetItem): Deleted.
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::frameID const):
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::createItemTree):
(WebCore::HistoryController::updateCurrentItem):
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::~WebBackForwardList):
(WebKit::WebBackForwardList::pageClosed):
(WebKit::WebBackForwardList::addItem):
(WebKit::WebBackForwardList::goToItem):
(WebKit::WebBackForwardList::goToProvisionalItem):
(WebKit::WebBackForwardList::goToItemInternal):
(WebKit::WebBackForwardList::clearProvisionalItem):
(WebKit::WebBackForwardList::currentItem const):
(WebKit::WebBackForwardList::backItem const):
(WebKit::WebBackForwardList::forwardItem const):
(WebKit::WebBackForwardList::itemAtIndex const):
(WebKit::WebBackForwardList::backListCount const):
(WebKit::WebBackForwardList::forwardListCount const):
(WebKit::WebBackForwardList::backListAsAPIArrayWithLimit const):
(WebKit::WebBackForwardList::forwardListAsAPIArrayWithLimit const):
(WebKit::WebBackForwardList::removeAllItems):
(WebKit::WebBackForwardList::clear):
(WebKit::WebBackForwardList::backForwardListState const):
(WebKit::WebBackForwardList::loggingString):
* Source/WebKit/UIProcess/WebBackForwardList.h:
(WebKit::WebBackForwardList::provisionalOrCurrentIndex):
(WebKit::WebBackForwardList::provisionalOrCurrentIndex const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::backForwardGoToProvisionalItem):
(WebKit::WebPageProxy::backForwardClearProvisionalItem):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::goToProvisionalItem):
(WebKit::WebBackForwardListProxy::clearProvisionalItem):
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::dumpHistoryItem):
* Source/WebKitLegacy/mac/History/BackForwardList.h:
* Source/WebKitLegacy/mac/History/BackForwardList.mm:
(BackForwardList::BackForwardList):
(BackForwardList::goToProvisionalItem):
(BackForwardList::clearProvisionalItem):
* Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
(-[WebHistoryItem description]):
* Source/WebKitLegacy/mac/History/WebHistoryItemPrivate.h:
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(dumpHistoryItem):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DecidePolicyForNavigationAction.mm:
(TEST(WebKit, DecidePolicyForNavigationActionGoBackAfterCancel)):

Canonical link: <a href="https://commits.webkit.org/285421@main">https://commits.webkit.org/285421@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4536a18c06d079c8baa712e3d6f397748b26e0ba

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72510 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51931 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76694 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23719 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74625 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59735 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23541 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57099 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15596 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75577 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47002 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62440 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37526 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43653 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19908 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22069 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65526 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20265 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78360 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16751 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19404 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65545 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16799 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62453 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64806 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13086 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6728 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11145 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47729 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2514 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48798 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50088 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48541 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebBackForwardList.h:
```diff
@@ -96,11 +99,17 @@ class WebBackForwardList : public API::ObjectImpl<API::Object::Type::BackForward
+    void goToItemInternal(WebBackForwardListItem&, std::optional<size_t>& indexToUpdate);
+
+    std::optional<size_t>& provisionalOrCurrentIndex() { return m_provisionalIndex ? m_provisionalIndex : m_currentIndex; }
+    std::optional<size_t> provisionalOrCurrentIndex() const { return m_provisionalIndex ? m_provisionalIndex : m_currentIndex; }
```
This looks like a mistake.  Could you make there be only one and be const and return a non-reference?  If you need a setter, make something that is more clearly a setter.

charliewolfe committed:
Clean up UI process back/forward state during provisional navigations
https://bugs.webkit.org/show_bug.cgi?id=281704
rdar://138150829

Reviewed by Alex Christensen.

When a back/forward navigation begins, we update the back/forward list in the UI process before the
navigation commits. This allows a client to quickly navigate back or forward again, even before the
navigation commits. Because of this, we have logic in `FrameLoader::continueLoadAfterNavigationPolicy`
to return the back/forward cursor to the previous item if the provisional navigation was canceled.
Instead of doing this, we should make the list in the UI process aware when a back/forward navigation has
started but not yet committed. The back/forward list will then use the provisional cursor until it is
either cleared or notified to switch to a non-provisional item.

`HistoryItem` now no longer needs to track whether it belongs to a navigated frame, so `m_isTargetItem`
can be removed.

I also added a test to ensure the back/forward cursor still returns to the previous item after canceling
a provisional navigation.

* LayoutTests/contentfiltering/allow-after-add-data-expected.txt:
* LayoutTests/contentfiltering/allow-after-finished-adding-data-expected.txt:
* LayoutTests/contentfiltering/allow-after-response-expected.txt:
* LayoutTests/contentfiltering/allow-after-will-send-request-expected.txt:
* LayoutTests/contentfiltering/allow-never-expected.txt:
* LayoutTests/contentfiltering/block-after-add-data-expected.txt:
* LayoutTests/contentfiltering/block-after-add-data-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-add-data-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-finished-adding-data-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-response-expected.txt:
* LayoutTests/contentfiltering/block-after-response-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-response-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-then-allow-unblock-expected.txt:
* LayoutTests/contentfiltering/block-after-will-send-request-then-deny-unblock-expected.txt:
* LayoutTests/contentfiltering/block-never-expected.txt:
* LayoutTests/fast/history/back-from-page-with-focused-iframe-expected.txt:
* LayoutTests/fast/history/form-submit-in-frame-expected.txt:
* LayoutTests/fast/history/form-submit-in-frame-via-onclick-expected.txt:
* LayoutTests/fast/history/gesture-before-onload-form-submit-expected.txt:
* LayoutTests/fast/history/gesture-before-onload-location-href-expected.txt:
* LayoutTests/fast/history/history-back-initial-vs-final-url-expected.txt:
* LayoutTests/fast/history/redirect-via-iframe-expected.txt:
* LayoutTests/fast/history/same-document-iframes-changing-fragment-expected.txt:
* LayoutTests/fast/history/same-document-iframes-changing-pushstate-expected.txt:
* LayoutTests/fast/loader/dynamic-iframe-extra-back-forward-item-expected.txt:
* LayoutTests/fast/loader/form-submission-before-load-get-expected.txt:
* LayoutTests/fast/loader/form-submission-before-load-post-expected.txt:
* LayoutTests/fast/loader/fragment-after-redirect-gets-back-entry-expected.txt:
* LayoutTests/fast/loader/frame-location-change-not-added-to-history-expected.txt:
* LayoutTests/fast/loader/frame-src-change-added-to-history-expected.txt:
* LayoutTests/fast/loader/frame-src-change-not-added-to-history-expected.txt:
* LayoutTests/fast/loader/stateobjects/pushstate-clears-forward-history-expected.txt:
* LayoutTests/fast/loader/subframe-navigate-during-main-frame-load-expected.txt:
* LayoutTests/http/tests/history/back-with-fragment-change-expected.txt:
* LayoutTests/http/tests/history/history-replace-updates-current-item-expected.txt:
* LayoutTests/http/tests/history/redirect-200-refresh-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-200-refresh-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-301-expected.txt:
* LayoutTests/http/tests/history/redirect-302-expected.txt:
* LayoutTests/http/tests/history/redirect-303-expected.txt:
* LayoutTests/http/tests/history/redirect-307-expected.txt:
* LayoutTests/http/tests/history/redirect-js-document-location-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-document-location-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-document-location-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-js-form-submit-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-form-submit-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-form-submit-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-assign-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-assign-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-assign-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-href-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-href-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-href-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-replace-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-replace-2-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-js-location-replace-before-load-expected.txt:
* LayoutTests/http/tests/history/redirect-meta-refresh-0-seconds-expected.txt:
* LayoutTests/http/tests/history/redirect-meta-refresh-2-seconds-expected.txt:
* LayoutTests/http/tests/navigation/anchor-basic-expected.txt:
* LayoutTests/http/tests/navigation/anchor-goback-expected.txt:
* LayoutTests/http/tests/navigation/anchor-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/back-to-slow-frame-expected.txt:
* LayoutTests/http/tests/navigation/back-twice-without-commit-expected.txt:
* LayoutTests/http/tests/navigation/document-location-click-expected.txt:
* LayoutTests/http/tests/navigation/document-location-click-timeout-expected.txt:
* LayoutTests/http/tests/navigation/document-location-mouseover-expected.txt:
* LayoutTests/http/tests/navigation/document-location-onload-expected.txt:
* LayoutTests/http/tests/navigation/document-location-script-expected.txt:
* LayoutTests/http/tests/navigation/dynamic-iframe-dynamic-form-back-entry-expected.txt:
* LayoutTests/http/tests/navigation/error404-frames-expected.txt:
* LayoutTests/http/tests/navigation/forward-and-cancel-expected.txt:
* LayoutTests/http/tests/navigation/javascriptlink-basic-expected.txt:
* LayoutTests/http/tests/navigation/javascriptlink-goback-expected.txt:
* LayoutTests/http/tests/navigation/javascriptlink-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/location-assign-adds-history-item-expected.txt:
* LayoutTests/http/tests/navigation/location-href-set-adds-history-item-expected.txt:
* LayoutTests/http/tests/navigation/location-replace-adds-history-item-expected.txt:
* LayoutTests/http/tests/navigation/location-set-adds-history-item-expected.txt:
* LayoutTests/http/tests/navigation/lockedhistory-iframe-expected.txt:
* LayoutTests/http/tests/navigation/metaredirect-basic-expected.txt:
* LayoutTests/http/tests/navigation/metaredirect-frames-expected.txt:
* LayoutTests/http/tests/navigation/metaredirect-goback-expected.txt:
* LayoutTests/http/tests/navigation/metaredirect-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/multiple-back-forward-entries-expected.txt:
* LayoutTests/http/tests/navigation/navigation-interrupted-by-fragment-expected.txt:
* LayoutTests/http/tests/navigation/new-window-redirect-history-expected.txt:
* LayoutTests/http/tests/navigation/onload-navigation-iframe-2-expected.txt:
* LayoutTests/http/tests/navigation/onload-navigation-iframe-expected.txt:
* LayoutTests/http/tests/navigation/onload-navigation-iframe-timeout-expected.txt:
* LayoutTests/http/tests/navigation/parsed-iframe-dynamic-form-back-entry-expected.txt:
* LayoutTests/http/tests/navigation/post-basic-expected.txt:
* LayoutTests/http/tests/navigation/post-frames-expected.txt:
* LayoutTests/http/tests/navigation/post-frames-goback1-expected.txt:
* LayoutTests/http/tests/navigation/post-frames-goback1-uncached-expected.txt:
* LayoutTests/http/tests/navigation/post-goback-same-url-expected.txt:
* LayoutTests/http/tests/navigation/post-goback1-expected.txt:
* LayoutTests/http/tests/navigation/post-goback2-expected.txt:
* LayoutTests/http/tests/navigation/postredirect-frames-expected.txt:
* LayoutTests/http/tests/navigation/postredirect-frames-goback1-expected.txt:
* LayoutTests/http/tests/navigation/postredirect-goback2-expected.txt:
* LayoutTests/http/tests/navigation/postredirect-reload-expected.txt:
* LayoutTests/http/tests/navigation/redirect-cycle-expected.txt:
* LayoutTests/http/tests/navigation/redirect-load-no-form-restoration-expected.txt:
* LayoutTests/http/tests/navigation/redirect-on-back-updates-history-item-expected.txt:
* LayoutTests/http/tests/navigation/redirect-on-reload-updates-history-item-expected.txt:
* LayoutTests/http/tests/navigation/redirect302-basic-expected.txt:
* LayoutTests/http/tests/navigation/redirect302-frames-expected.txt:
* LayoutTests/http/tests/navigation/redirect302-goback-expected.txt:
* LayoutTests/http/tests/navigation/redirect302-metaredirect-expected.txt:
* LayoutTests/http/tests/navigation/redirect302-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/relativeanchor-basic-expected.txt:
* LayoutTests/http/tests/navigation/relativeanchor-frames-expected.txt:
* LayoutTests/http/tests/navigation/relativeanchor-goback-expected.txt:
* LayoutTests/http/tests/navigation/response204-expected.txt:
* LayoutTests/http/tests/navigation/restore-form-state-https-expected.txt:
* LayoutTests/http/tests/navigation/sec-fetch-site-header-expected.txt:
* LayoutTests/http/tests/navigation/sec-fetch-site-header-on-crossorigin-redirection-expected.txt:
* LayoutTests/http/tests/navigation/sec-fetch-site-header-on-redirection-expected.txt:
* LayoutTests/http/tests/navigation/slowmetaredirect-basic-expected.txt:
* LayoutTests/http/tests/navigation/slowtimerredirect-basic-expected.txt:
* LayoutTests/http/tests/navigation/statistics-expected.txt:
* LayoutTests/http/tests/navigation/success200-basic-expected.txt:
* LayoutTests/http/tests/navigation/success200-frames-expected.txt:
* LayoutTests/http/tests/navigation/success200-frames-loadsame-expected.txt:
* LayoutTests/http/tests/navigation/success200-goback-expected.txt:
* LayoutTests/http/tests/navigation/success200-loadsame-expected.txt:
* LayoutTests/http/tests/navigation/success200-reload-expected.txt:
* LayoutTests/http/tests/navigation/success200-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/timerredirect-basic-expected.txt:
* LayoutTests/http/tests/navigation/timerredirect-frames-expected.txt:
* LayoutTests/http/tests/navigation/timerredirect-goback-expected.txt:
* LayoutTests/http/tests/navigation/timerredirect-subframeload-expected.txt:
* LayoutTests/http/tests/navigation/window-open-adds-history-item-expected.txt:
* LayoutTests/http/tests/navigation/window-open-adds-history-item2-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-allowed-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-default-ignored-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-get-allowed-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-get-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-javascript-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/1.1/form-action-src-self-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/upgrade-insecure-requests/proper-uir-on-navigation-expected.txt:
* LayoutTests/platform/glib/http/tests/navigation/error404-basic-expected.txt:
* LayoutTests/platform/glib/http/tests/navigation/error404-goback-expected.txt:
* LayoutTests/platform/glib/http/tests/navigation/postredirect-basic-expected.txt:
* LayoutTests/platform/glib/http/tests/navigation/postredirect-goback1-expected.txt:
* LayoutTests/platform/gtk/http/tests/navigation/error404-subframeload-expected.txt:
* LayoutTests/platform/gtk/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/error404-basic-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/error404-goback-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/error404-subframeload-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/postredirect-basic-expected.txt:
* LayoutTests/platform/ios/http/tests/navigation/postredirect-goback1-expected.txt:
* LayoutTests/platform/mac-ventura/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/mac-wk1/http/tests/navigation/sec-fetch-site-header-expected.txt:
* LayoutTests/platform/mac-wk1/http/tests/navigation/sec-fetch-site-header-on-crossorigin-redirection-expected.txt:
* LayoutTests/platform/mac-wk1/http/tests/navigation/sec-fetch-site-header-on-redirection-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/error404-basic-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/error404-goback-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/error404-subframeload-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/javascriptlink-frames-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/postredirect-basic-expected.txt:
* LayoutTests/platform/mac/http/tests/navigation/postredirect-goback1-expected.txt:
* LayoutTests/platform/wpe/http/tests/navigation/javascriptlink-frames-expected.txt:
* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::setProvisionalItem):
(WebCore::BackForwardController::clearProvisionalItem):
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::reset):
(WebCore::HistoryItem::setChildItem):
(WebCore::HistoryItem::isTargetItem const): Deleted.
(WebCore::HistoryItem::setIsTargetItem): Deleted.
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::frameID const):
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::createItemTree):
(WebCore::HistoryController::updateCurrentItem):
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::~WebBackForwardList):
(WebKit::WebBackForwardList::pageClosed):
(WebKit::WebBackForwardList::addItem):
(WebKit::WebBackForwardList::goToItem):
(WebKit::WebBackForwardList::goToProvisionalItem):
(WebKit::WebBackForwardList::goToItemInternal):
(WebKit::WebBackForwardList::clearProvisionalItem):
(WebKit::WebBackForwardList::currentItem const):
(WebKit::WebBackForwardList::backItem const):
(WebKit::WebBackForwardList::forwardItem const):
(WebKit::WebBackForwardList::itemAtIndex const):
(WebKit::WebBackForwardList::backListCount const):
(WebKit::WebBackForwardList::forwardListCount const):
(WebKit::WebBackForwardList::backListAsAPIArrayWithLimit const):
(WebKit::WebBackForwardList::forwardListAsAPIArrayWithLimit const):
(WebKit::WebBackForwardList::removeAllItems):
(WebKit::WebBackForwardList::clear):
(WebKit::WebBackForwardList::backForwardListState const):
(WebKit::WebBackForwardList::loggingString):
* Source/WebKit/UIProcess/WebBackForwardList.h:
(WebKit::WebBackForwardList::provisionalOrCurrentIndex):
(WebKit::WebBackForwardList::provisionalOrCurrentIndex const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::backForwardGoToProvisionalItem):
(WebKit::WebPageProxy::backForwardClearProvisionalItem):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::goToProvisionalItem):
(WebKit::WebBackForwardListProxy::clearProvisionalItem):
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::dumpHistoryItem):
* Source/WebKitLegacy/mac/History/BackForwardList.h:
* Source/WebKitLegacy/mac/History/BackForwardList.mm:
(BackForwardList::BackForwardList):
(BackForwardList::goToProvisionalItem):
(BackForwardList::clearProvisionalItem):
* Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
(-[WebHistoryItem description]):
* Source/WebKitLegacy/mac/History/WebHistoryItemPrivate.h:
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(dumpHistoryItem):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DecidePolicyForNavigationAction.mm:
(TEST(WebKit, DecidePolicyForNavigationActionGoBackAfterCancel)):

Canonical link: https://commits.webkit.org/285421@main
